File: SDE_sheet/Arrays/11_missing_and_repeating_number/solution.java
public int[] findMissingRepeatingNumbers(int[] a) {
    int n = a.length;
    int xr = 0;
    for (int i = 0; i < n; i++) {
        xr ^= a[i];
        xr ^= (i + 1);
    }
    int number = (xr & ~(xr - 1));
    int zero = 0, one = 0;
    for (int i = 0; i < n; i++) {
        if ((a[i] & number) != 0) {
            one ^= a[i];
        } else {
            zero ^= a[i];
        }
    }
    for (int i = 1; i <= n; i++) {
        if ((i & number) != 0) {
            one ^= i;
        } else {
            zero ^= i;
        }
    }
    for (int i = 0; i < n; i++) {
        if (a[i] == zero) {
            return new int[] {zero, one};
        }
    }
    return new int[] {one, zero};
}

File: SDE_sheet/Arrays/11_missing_and_repeating_number/solution.rs
fn find_missing_repeating_numbers(a: Vec<i32>) -> (i32, i32) {
    let n = a.len();
    let mut xr = 0;
    for i in 0..n {
        xr ^= a[i];
        xr ^= (i + 1) as i32;
    }
    let number = xr & !(xr - 1);
    let mut zero = 0;
    let mut one = 0;
    for i in 0..n {
        if a[i] & number != 0 {
            one ^= a[i];
        } else {
            zero ^= a[i];
        }
    }
    for i in 1..=n {
        if (i as i32) & number != 0 {
            one ^= i as i32;
        } else {
            zero ^= i as i32;
        }
    }
    for i in 0..n {
        if a[i] == zero {
            return (zero, one);
        }
    }
    (one, zero)
}

